packages/optimizely-sdk/lib/core/audience_evaluator/index.js

```
@@ -13,34 +13,47 @@
+   *                                          Examples: ["5", "6"], ["and", ["or", "1", "2"], "3"]
+   * @param  {Object}   audiencesById         Object providing access to full audience objects for audience IDs
+   *                                          contained in audienceConditions. Keys should be audience IDs, values
+   *                                          should be full audience objects with conditions properties
+   * @param  {Object}   [userAttributes]      User attributes which will be used in determining if audience conditions
+   *                                          are met. If not provided, defaults to an empty object
+   * @return {Boolean}                        true if the user attributes match the given audience conditions, false
+   *                                          otherwise
    */
-  evaluate: function(audiences, userAttributes) {
+  evaluate: function(audienceConditions, audiencesById, userAttributes) {
     // if there are no audiences, return true because that means ALL users are included in the experiment
-    if (!audiences || audiences.length === 0) {
+    if (!audienceConditions || audienceConditions.length === 0) {
       return true;
     }
 
     if (!userAttributes) {
       userAttributes = {};
     }
 
-    for (var i = 0; i < audiences.length; i++) {
-      var audience = audiences[i];
-      var conditions = audience.conditions;
-      if (conditionEvaluator.evaluate(conditions, userAttributes)) {
-        return true;
+    var evaluateConditionWithUserAttributes = function(condition) {
+      return customAttributeConditionEvaluator.evaluate(condition, userAttributes);
+    };
+
+    var evaluateAudience = function(audienceId) {
+      var audience = audiencesById[audienceId];
+      if (audience) {
+        return conditionTreeEvaluator.evaluate(audience.conditions, evaluateConditionWithUserAttributes);
       }
+      return null;
```

@nchilada @mikeng13 Any thoughts on returning `null` here? Maybe this should throw an error, and project config should reject a datafile as invalid if an experiment has invalid audience ids.

---

I believe returning `null` is correct here - this function is being called at activate time so it's probably too late to decide that the datafile is invalid and to stop handling API calls.

That said, we could add an additional check for invalid audience references during initialization. If we do so, we reduce this `return null` to a mere sanity check.

[In the design doc](https://docs.google.com/document/d/12aoOd1g5naAynHGdMW1vpaGbBaaUCQHGtRhUDfFRGOg/edit#heading=h.by7wosg2x56t) I'd written:

> If an audience can't be found in [`audiences` or `typedAudiences`], the SDK may fail gracefully or spectacularly, during initialization or during experiment activation—whatever it was doing before.

but I don't mind if someone want to propose that we consistently check for invalid audience references during initialization. @mikeng13 @thomaszurkan-optimizely any thoughts on this?